### PR TITLE
LSTM and Bidirectional layers

### PR DESCRIPTION
**Pull request by Gaarv**
## What changes were proposed in this pull request?

* added LSTM and Bidirectional layers
* moved identical implementations from Sequential/NeuralNet to Model trait, override not necessary
* deleted optimizer.scala as not used anymore following PR #45
* added a "SequenceClassification" example using a Bidirectional LSTM. From an example found back in the days learning Keras. Not sure if really useful, I used it mainly to validate ScalNet results.
* added a "SequenceGenerator" util in test used for above example. Same comment applies.

## How was this patch tested?

* fairly basic unit tests for LSTM and Bidirectional since they don't have much parameters. I hesitated to make tests for different modes (ADD, CONCAT, etc) but those are already covered in DL4J and maybe not really needed at a wrapper level.
* see example "SequenceClassification" in examples/dl4j/recurrent
* don't pay attention to failed travis builds on my end, which get ok when restarting later on :/ Not sure why, might be related to when snapshots are getting build on sonatype (time of the day).

## Needs review

* in Bidirectional ScalNet layer, inputShape and outputShape have been set to List.empty since this is a wrapper layer. Not sure between this and maybe "layer.inputShape", layer being the LSTM passed in constructor ?
* contrary to other layers, used the constructor rather than Builder since it was publicly available

### CODE REVIEW

File: src/main/scala/org/deeplearning4j/scalnet/models/Sequential.scala [hunk 1]
```diff
@@ -101,22 +93,6 @@ class Sequential(miniBatch: Boolean, biasInit: Double, rngSeed: Long) extends Mo
     model.init()
   }
 
-  override def fit(iter: DataSetIterator, nbEpoch: Int, listeners: List[IterationListener]): Unit = {
-    model.setListeners(listeners.asJavaCollection)
-    for (epoch <- 0 until nbEpoch) {
-      logger.info("Epoch " + epoch)
```
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/models/Sequential.scala]: 👍 wanted to do this for a long time

File: src/main/scala/org/deeplearning4j/scalnet/layers/recurrent/Bidirectional.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.deeplearning4j.nn.conf.layers
+import org.deeplearning4j.nn.conf.layers.recurrent.Bidirectional.Mode
+import org.deeplearning4j.scalnet.layers.core.Layer
+
+class Bidirectional(layer: Layer, mode: Mode, override val name: String = "") extends Layer {
+
+  override def compile: layers.Layer = new layers.recurrent.Bidirectional(mode, layer.compile)
+
+  override def inputShape: List[Int] = List.empty
```
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/layers/recurrent/Bidirectional.scala]: We can do this analogous to DL4J and e.g. use a `LayerWrapper` class. A wrapper has a `Layer` field called `underlying` and shape info is redirected to the underlying layer. For instance:

```scala
...
val underlying = layerFromConstructor
override der inputShape: List[Int] = underlying.inputShape
```

and then in a wrapper you just set the underlying, not the shapes. something like that.

Comment (Gaarv) [reply src/main/scala/org/deeplearning4j/scalnet/layers/recurrent/Bidirectional.scala]: Alright thanks :) will look into it !

**Comment by maxpumperla** [comment]
thanks @Gaarv

**Pull request merged**